### PR TITLE
Fix errors='ignore' being ignored in astype #30324

### LINKED ISSUE #30324: errors='ignore' does not work on df.replace({col : type}) if col not found
This may be considered a feature request

#### Code Sample, a copy-pastable example if possible

```python
df = pd.DataFrame({'a' : [1, 2, 3]})
df.astype({'b' : str}, errors='ignore')

KeyError: 'Only a column name can be used for the key in a dtype mappings argument.'

```
#### Problem description

A KeyError is still raised if the column name is not found. I'd think errors='ignore' should suppress this error.

#### Expected Output

Unchanged dataframe.

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.3.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 17.7.0
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : en_US.UTF-8
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 0.25.2
numpy            : 1.17.2
pytz             : 2019.3
dateutil         : 2.8.0
pip              : 19.3.1
setuptools       : 41.4.0
Cython           : 0.29.13
pytest           : 5.2.1
hypothesis       : None
sphinx           : 2.2.0
blosc            : None
feather          : None
xlsxwriter       : 1.2.2
lxml.etree       : 4.4.1
html5lib         : 1.0.1
pymysql          : None
psycopg2         : None
jinja2           : 2.10.3
IPython          : 7.8.0
pandas_datareader: None
bs4              : 4.8.1
bottleneck       : 1.2.1
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.4.1
matplotlib       : 3.1.1
numexpr          : 2.7.0
odfpy            : None
openpyxl         : 3.0.0
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : 1.3.1
sqlalchemy       : 1.3.10
tables           : 3.5.2
xarray           : None
xlrd             : 1.2.0
xlwt             : 1.3.0
xlsxwriter       : 1.2.2
</details>

### PR DESCRIPTION
**Pull request by naomi172839**
- [x] closes #30324 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Review by WillAyd — CHANGES_REQUESTED** [review]
Can you add test(s)? Should be the first part of any PR

**Comment by naomi172839** [comment]
@WillAyd I added unit tests to test_generic.py.  I have never used pytest before and am having trouble building the c components to test it myself.  Hopefully it was done right.

File: pandas/tests/generic/test_generic.py [hunk 1]
```diff
@@ -948,3 +947,21 @@ def test_axis_classmethods(self, box):
             assert obj._get_axis_number(v) == box._get_axis_number(v)
             assert obj._get_axis_name(v) == box._get_axis_name(v)
             assert obj._get_block_manager_axis(v) == box._get_block_manager_axis(v)
+
+    # Tests that the ignore keyword causes the intended functionality
+    def test_astype_ignores(self):
```
Comment (alimcmaster1) [inline pandas/tests/generic/test_generic.py]: Nitpick: you might want to consider pytest.mark.parametrize
https://docs.pytest.org/en/latest/parametrize.html#pytest-mark-parametrize

You can use `from contextlib import ExitStack as does_not_raise` as a context manager for the final case as per http://doc.pytest.org/en/latest/example/parametrize.html#parametrizing-conditional-raising

File: pandas/tests/generic/test_generic.py [hunk 2]
```diff
@@ -948,3 +947,21 @@ def test_axis_classmethods(self, box):
             assert obj._get_axis_number(v) == box._get_axis_number(v)
             assert obj._get_axis_name(v) == box._get_axis_name(v)
             assert obj._get_block_manager_axis(v) == box._get_block_manager_axis(v)
+
```
Comment (jreback) [inline pandas/tests/generic/test_generic.py]: this needs to go in pandas/tests/frame/test_dtype with the other astype tests.

add similar for series (or does this not affect series?)

File: pandas/tests/generic/test_generic.py [hunk 3]
```diff
@@ -948,3 +947,21 @@ def test_axis_classmethods(self, box):
...
+    # Tests that the ignore keyword causes the intended functionality
```
Comment (jreback) [inline pandas/tests/generic/test_generic.py]: comments go in side the test; add the issue number as a comment

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/generic/test_generic.py (hunk 1 above)
Comment (naomi172839) [reply pandas/tests/generic/test_generic.py]: I would love to learn more about pytest.mark.parametrize in the future but I'm not super confident implementing it here without trying it out in other smaller/less important projects first.  If you think it is absolutely essential then I will push another commit once I figure it out.

File: pandas/tests/generic/test_generic.py (hunk 2 above)
Comment (naomi172839) [reply pandas/tests/generic/test_generic.py]: Moved

Comment (naomi172839) [reply pandas/tests/generic/test_generic.py]: Series testing seems to be unnecessary as an error is thrown by another method prior to the astype method getting a change to throw one.

File: pandas/tests/generic/test_generic.py (hunk 3 above)
Comment (naomi172839) [reply pandas/tests/generic/test_generic.py]: Done.

**Comment by naomi172839** [comment]
I seemed to have messed up the GIT... Let me know if I need to start over:/

**Comment by alimcmaster1** [comment]
> I seemed to have messed up the GIT... Let me know if I need to start over:/

You can still use this PR/branch - fix it up locally and use —force if needed

**Comment by naomi172839** [comment]
OK so I fixed the git.  The tests actually seemed to already exist to make sure that the error was thrown.  I am not sure why it did not fail before.

**Comment by jreback** [comment]
can you add a whatsnew note

**Comment by naomi172839** [comment]
@jreback When I add my own whatsnew entry it seems to create a conflict that I am unsure how to resolve.  I already updated my local whatsnew file to the latest version.

**Comment by TomAugspurger** [comment]
Hmm, looking back at the original issue, I'm not sure if this is what we want to do.

The `errors` keyword is used to handle raising when there's in issue with the conversion of values. This PR overloads `errors` to handle raising when there are non-conversion issues (incorrect keys in `dtype` dict for example). I don't think we want to overload the meaning of `errors` like that.

**Comment by TomAugspurger** [comment]
Apologies for not catching that earlier @naomi172839.

**Comment by naomi172839** [comment]
@tomauspurger would you then suggest adding an additional parameter to the astype method to allow the user to ignore the KeyErrors?

**Comment by TomAugspurger** [comment]
That's one option, though I'm not sure I'd find it useful enough. Others may though. Perhaps @climatebrad can share his original usecase.

**Comment by climatebrad** [comment]
My [usecase](https://github.com/climatebrad/stop-and-frisk/) is that I was joining several dataframes generated from different years of NYPD stop-and-frisk datasets into one multiyear dataframe. A lot of the work to keep the dataframes manageable was making sure that categorical data was saved as such. So I had a list of all of the columns that needed to be categorical data and was setting them. (Technically I converted to object type, did more processing, then converted to category type.) This code failed:

```
cat_set = set(CLEAN_CAT_VALUES) | set(CAT_FILL_NA_VALUES) 
data = data.astype({cat : 'object' for cat in cat_set}, errors='ignore')
```

Here's the diff where I dealt with what I consider the incorrect `errors='ignore'` behavior. (I had to explicitly exclude columns that didn't exist in the dataframe I was editing.) Not the hardest workaround, but I do still think that `errors='ignore'` should, you know, ignore errors.

https://github.com/climatebrad/stop-and-frisk/commit/d755084c100fd0f8f8d883f9f9a471ddbb0a5c02#diff-a0ebe5d9008a8b884ea03629345f9735R178

```
cat_set = set(CLEAN_CAT_VALUES) | set(CAT_FILL_NA_VALUES) 
cat_set = cat_set.intersection(set(data.columns))
data = data.astype({cat : 'object' for cat in cat_set}, errors='ignore')
```

**Comment by naomi172839** [comment]
@climatebrad I do agree that errors='ignore' at least implies that errors will be ignored but @tomaugspurger does have a point.  It would overload the method and that's probably not the best solution.

**Comment by WillAyd** [comment]
Looks like the 1.0 whatsnew got mangled in this - can you check that out from master?

**Comment by naomi172839** [comment]
Sorry that it took so long, I had to figure out how to check out a single file.

**Comment by WillAyd** [comment]
@naomi172839 not sure why the whatsnew files are collected here, but can you try the following on your branch?

```python
git fetch upstream/master
git checkout upstream/master pandas/doc/source/whatsnew/v1*
```

Then re-push? I think should fix the issues here (can add a whatsnew note after we fix these up)

**Comment by naomi172839** [comment]
@WillAyd So I tried that with no real luck.  I tried deleting the files and rechecking them out which got me to where we are now.  I dont have high hopes.

**Comment by WillAyd** [comment]
Your last push looks a lot better. The only outstanding issue is that the file permissions have changed for the v1.0.0 whatsnew.

If you do `git checkout upstream/master pandas/doc/source/whatsnew/v1.0.0.rst` now does `git status` show anything as staged?

**Comment by naomi172839** [comment]
Output is below:
nbonnin@Naomis-MacBook-Pro pandas % git checkout upstream/master doc/source/whatsnew/v1.0.0.rst
Updated 0 paths from 247718065
nbonnin@Naomis-MacBook-Pro pandas % git status
On branch #30324
Your branch is up to date with 'origin/#30324'.

nothing to commit, working tree clean

So I don't think that there is anything staged

**Comment by WillAyd** [comment]
Hmm OK. I guess git checkout might not grab file permissions.

Since you are on macOS you can just do `chmod 1644 pandas/doc/source/whatsnew/v1.0.0.rst` instead to restore back to where it was

**Comment by naomi172839** [comment]
Done, though its not really showing any changes and zero differences from upstream/master for that file

**Comment by WillAyd** [comment]
Is your `upstream` pointing to pandas? If not sure, can you post the output of `git remote -v show`?

**Comment by naomi172839** [comment]
Output 
nbonnin@Naomis-MacBook-Pro pandas % git remote -v show
origin  https://github.com/naomi172839/pandas.git (fetch)
origin  https://github.com/naomi172839/pandas.git (push)
upstream        https://github.com/pandas-dev/pandas.git (fetch)
upstream        https://github.com/pandas-dev/pandas.git (push)

**Comment by WillAyd** [comment]
Just for posterity try this one more time - it worked for the other whatsnew files so not sure what makes the 1.0.0 any different but maybe the changed permissions prevented this from working

```sh
git fetch upstream/master
git checkout upstream/master pandas/doc/source/whatsnew/v1.0.0.rst
```

**Comment by naomi172839** [comment]
Doing that, git thought that there were changes so I went ahead and pushed it.

**Comment by WillAyd** [comment]
Hmm still didn't work. If you run `git diff upstream/master --name-only` locally none of the whatsnew files should show, but they seem to here on GitHub

If all else fails maybe just download the file directly from GitHub

**Comment by naomi172839** [comment]
Output:
...
doc/source/whatsnew/v0.21.0.rst
doc/source/whatsnew/v0.23.0.rst
doc/source/whatsnew/v0.25.0.rst
doc/source/whatsnew/v0.25.3.rst
doc/source/whatsnew/v1.0.1.rst
doc/source/whatsnew/v1.0.2.rst
doc/source/whatsnew/v1.1.0.rst
...

v1.0.0.rst isn't listed.  Nevertheless I tried downloading straight from github.  It still seems to have made no difference.

**Comment by WillAyd** [comment]
Hmm OK. Actually just needed to do a `git merge upstream/master` and repush

I did that to fix things up here, so just `git pull` from your local branch to grab the changes and should be able to move forward

**Comment by naomi172839** [comment]
Ok great!  that did alot.  At this point it should be just adding the what's new message correct?  I'm not super great at git (obviously).  Just want to make sure before I push another commit.

**Comment by WillAyd** [comment]
Hmm just reading back through the comment history I do agree with @TomAugspurger that this blurs the purpose of the `errors` keyword; any chance you thought about alternative implementations?

**Comment by naomi172839** [comment]
The only real alternative that I came up with was to add another optional argument to the method, but I am not sure that that is really the best solution.

Looking into it now, it could also be an option like mode.sim_interactive is.  I don't really know if that is any better though.

**Comment by TomAugspurger** [comment]
I think I'm still against adding this change. I think the errors keyword should just be for controlling the values conversion.

If we want to allow unused keys to be provided, we can add a keyword dedicated for that. There are other context like `.rename` where people have expressed a desire for extra / unused keys to raise (they're currently ignored in rename).


```python
df = pd.DataFrame({"A": [1, 2]})
df.astype({"A": "int", "B": "float"}, ignore_unused_keys=True)

df.rename(columns={"A": "a", "B", "b"}, ignore_unused_keys=False)  # raises
```

**Comment by climatebrad** [comment]
Although I am in the camp of "errors='ignore'" should mean "ignore errors" I'd be perfectly happy with a separate keyword to allow for this functionality. 

I'd make it consistent with the `errors` syntax though. Here is language consistent with the `df.rename` documentation.

**missing_cols : {'raise', 'ignore'}, default 'raise'**
If ‘raise’, raise a KeyError when a dict of columns contains labels that are not present. If ‘ignore’, existing keys will be retyped and extra keys will be ignored.

**Comment by TomAugspurger** [comment]
Your proposal should be `extra_columns: {'raise', 'ignore'}`, or unused_columns, right? Not "missing"?

**Comment by WillAyd** [comment]
> Personally, I feel that ignore_unused_keys is a more descriptive argument. If everyone thinks this is a good/acceptable route to go, I would be happy to work on it.

Yea I think that would be clearer

**Comment by WillAyd** [comment]
@naomi172839 still active? Want to try to incorporate latest feedback?

**Comment by simonjayhawkins** [comment]
@naomi172839 closing as stale. ping if you want to continue.

**Pull request closed without merging**